feat(web): delayed modipress completion 🐵

jahorton opened the pull request:
This aims to cover a few "fun" edge cases that can easily arise during quick typing that involves modipresses:

1. For longpresses triggered during a modipress that has since been released, the underlying layer for the OSK will not revert until subkey-selection is complete.
    - It felt "weird" when it would swap back early, underneath the subkey menu.
    - Granted, it also feels weird to me when it stays in place, but it's a better representation of the system's state.
2. For _flicks_ triggered during a modipress that has since been released, the underlying layer for the OSK will not revert until the flick is complete.
    - Note:  flick completion may be triggered if a new, incoming tap is received.  In such a case:
        1. The flick completes as normal, for the original key from the modipressed layer.
        2. The modipress is then fully resolved, including reversion of the layer to the one that was originally active before the modipress
        3. And _then_ the new, incoming tap is processed - for the pre-modipress layer.

The main idea behind point 2 above:  during quick typing, it's very natural to lift the modipress finger to prepare to type a new key.  This could easily trigger earlier than intended, so we keep the modipress active while the flick is still active.  

Point 1 above then provides consistency with point 2 - in both cases, the layer "sticks" until the pending gesture is completed.

In case the user accidentally presses a new key either slightly before or simultaneously with release during this state, we can still allow the flick to complete.  Since this is generally during quick typing, and the user _did_ lift the modipress key, they've likely has anticipated the layer to return - and so we adjust for that expectation.

I will say that point 2 above was an interesting niche case to implement - it involved a second coming of the TWo caps problem (#7173 / #9802) from a slightly different angle.  While it does add some complexity, it should help make flicks more robust to certain natural 'optimizations' we often subconsciously make while typing quickly.

## Notes for future revisitation

❗ ❗ ❗

I ended up throwing in quite a number of other fixes into this one, as it held the big user test suite and was used as the anchor-point for demo builds.  Not exactly ideal, but it is what it is at this point.  FWIW, I did aim to keep the commits reasonably clean, so going commit by commit might help... though there was a merge from the base (including a master update) that may make certain aspects interesting to follow.

## User Testing

Note that most of these tests will be repeats of user tests from previous PRs.  Consider this largely a feature-branch regression-test suite...  and thus tests worthy of consideration for becoming _permanent_ regression tests once the feature-branch lands.

**TEST_10_KEY_ROTA**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "English...", then "Classic 10-key".
2. Verify that all of the keys numbered 1-9 aside from 7 have three or more lowercase letters as key hints.
3. Tap the `8` key once, then wait for a second.
4. Tap the `8` key twice in quick succession, then wait for a second.
    - Expected output: `8a`.
5. Tap the `8` key continuously for at least 2 seconds.
    - it should rotate through: `8a8`, `8aa`, `8ab`, `8ac`, `8aA`, `8aB`, `8aC`, then back to `8a8` (and continue from there)
    - stop tapping when the new character is a capital letter.
6. Verify that the key hints are now capitalized.
7. Double-tap the `9` key once.  A `D` should be emitted as the result.
8. Repeatedly tap `9` for at least five seconds, paying attention to the key hints.
    - Whenever a lowercase letter is the result, the key hints should be lowercase
    - Whenever an uppercase letter is the result, the key hints should be uppercase

**TEST_10_KEY_DIACRITICS**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select the text-area element - the first input-accepting element on the page.
    - Placeholder text:  "Type here".
    - Is directly underneath "Type in your language in this text area:"
2. Select "English...", then "Diacritic 10-key Rota". 
3. Tap the '◌̀' key.  (The key between backspace [above] and enter [below], representing a diacritic)
4. Double-tap the `8` key.
    - Expected result: `à`
5. Triple-tap the '◌̀' key - a circumflex diacritic should appear over the `à`.
6. Triple-tap the `9` key.
    - Expected total result: `àê`
7. Tap the '◌̀' key.  A diacritic should appear over the `ê`.
8. Deselect the text-area element.
9. Reselect the text-area element, placing the point of text entry (the caret) at the end, after the `ê`.
    - Note for future test-maintenance:  the point is to trigger a context-reset.
10. Double-tap the `8` key.
    - Expected final result:  `àềa`
        ![image](https://github.com/keymanapp/keyman/assets/25213402/de257a64-a7b8-4d12-8bd8-012642cd532e)
    - For clarity:
        - With a copy of the `à`'s accent-mark on top the original `ê`, as in the cropped screenshot above.  
            - It looks different in raw-text form here due to GitHub's font / rendering.
        - Note:  there is no matching accent-mark on the second `a`.

**TEST_APP_10_KEY_DIACRITICS**:  Using the Keyman for Android test artifact...

1. Download this KMP: https://jahorton.github.io/diacritic_rota.kmp
    - Can also be found via the main page at https://jahorton.github.io.
2. Open / install it for use within the Keyman app.
3. Tap the 7 key twice; verify that `77` is output.
4. Tap the '◌̀' key.  (The key between backspace [above] and enter [below], representing a diacritic)
5. Double-tap the `8` key.
    - Expected total result: `77à`
6. Triple-tap the '◌̀' key - a circumflex diacritic should appear over the `à`.
7. Triple-tap the `9` key.
    - Expected total result: `77àê`
8. Tap the '◌̀' key.  A diacritic should appear over the `ê`.
9. Open the in-app Settings menu.
    - Note, for test maintenance:  the point of this step and the next one is to trigger a context-reset.
10. Close the in-app Settings menu.
11. Double-tap the `8` key.
    - Expected final result:  `77àềa`
        ![image](https://github.com/keymanapp/keyman/assets/25213402/de257a64-a7b8-4d12-8bd8-012642cd532e)
    - For clarity:
        - With a copy of the `à`'s accent-mark on top the original `ê`, as in the cropped screenshot above.  
            - It looks different in raw-text form here due to GitHub's font / rendering.
        - Note:  there is no matching accent-mark on the second `a`.

**TEST_BASIC_SIMPLE_SHIFT**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "Norwegian" (for the SIL EuroLatin keyboard)
2. Tap the spacebar.  (The keyboard should be on the default [lowercase] layer after this.)
3. Tap the shift key once.
4. Verify that the keyboard changes to, and stays on, the shift layer.

**TEST_BASIC_MODIPRESS**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "Norwegian" (for the SIL EuroLatin keyboard)
2. Tap the spacebar.  (The keyboard should be on the default [lowercase] layer after this.)
3. Tap and hold the shift key.
4. Verify that the keyboard changes to the shift layer.
5. Tap the `A` key, verifying that it produces an `A`.
6. Release the shift key.
7. Verify that the keyboard changes to the default layer as a result of step 6.
8. Repeat steps 3 through 7 _quickly_, holding shift just long enough to tap the `shift` layer's `A` key, then releasing quickly.

**TEST_BASIC_MODIPRESS_HOLD**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "Norwegian" (for the SIL EuroLatin keyboard)
2. Tap the spacebar.  (The keyboard should be on the default [lowercase] layer after this.)
3. Tap and hold the shift key.
4. Verify that the keyboard changes to the shift layer.
5. Wait at least one second.
6. Release the shift key.
7. Verify that the keyboard changes to the default layer as a result of step 6.

**TEST_NUMERIC_FROM_SHIFT**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "Norwegian" (for the SIL EuroLatin keyboard)
2. Tap the spacebar.  (The keyboard should be on the default [lowercase] layer after this.)
3. Tap the shift key, making the shift layer active.
4. Tap and hold the numeric key (`123`) underneath the shift key.
5. Verify that the keyboard changes to the numeric layer - numbers in the top row, mathematical symbols elsewhere.
6. Longpress the `%` key and select the `‱` subkey.
7. Verify that `‱` is emitted as text.
8. Release the numeric key.
11. Verify that the keyboard changes back to the _shift_ layer.
12. Tap the numeric key, releasing it immediately.
13. Repeat steps 6 and 7, longpressing the `%` key and select the `‱` subkey.
14. Verify that the keyboard automatically changes back to the _default_ (lowercase) layer.

**TEST_DELAYED_SUBKEY**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "Norwegian" (for the SIL EuroLatin keyboard)
2. Tap the spacebar.  (The keyboard should be on the default [lowercase] layer after this.)
3. Tap the shift key, making the shift layer active.
4. Tap and hold the numeric key (`123`) underneath the shift key.
5. Verify that the keyboard changes to the numeric layer - numbers in the top row, mathematical symbols elsewhere.
6. Longpress the `%` key, but do not select a subkey yet.
7. Release the numeric key.
8. Verify that the keyboard changes _remains on the numeric layer_, with the subkey menu still displayed.
10. Select the `‱` subkey.
12. Verify that `‱` is emitted as text and that the layer has changed back to the _`shift`_ layer.

**TEST_DOUBLETAP_CAPS**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "Norwegian" - you should then see "Norwegian - EuroLatin (SIL)" in the space bar.
2. Type `All `.
3. Double-tap the shift key.  Afterward, the shift key should be highlighted, and its up-arrow should have a thin line underneath, as if underlining it.
4. Without touching the shift key at any point, type `CAPS `.
    - If this is impossible - i.e., the layer shifted out from underneath you - FAIL the test.
5. Single-tap the shift key.  Afterward, you should be returned to the base layer.
6. Type `work `.
7. Tap the shift key _once_.
8. Without touching the shift key at any point, type `Well`.
    - If this is impossible - i.e., the layer did not drop the the base layer when expected - FAIL the test.
    - Expected final result, in total: `All CAPS work Well`

**TEST_CUSTOM_MULTITAP_MODIFIER**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "English...", then "Gesture Prototyping".
2. Tap and hold the `◌́` key.  (The key between the spacebar [left] and the Enter key [right].)
3. A number of letters, including all of the vowels, should change to versions using the `◌́` accent mark.
4. Tap the `á` and verify that the corresponding letter is output.
5. Release the `◌́` key; you should be returned to the default layer.
6. Note the hints on the `◌́`, then triple-tap it, holding on the third tap.
7. A number of letters, including all of the vowels, should change to versions using the `◌̂` mark
    - This corresponds to the second diacritic from the `◌́`-key's hint.
8. Tap the `â` and verify that the corresponding letter is output.
9. Release the `◌́` key; you should be returned to the default layer.

TEST_ALTERNATING_SHIFT_AND_KEY:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "English...", then "Gesture Prototyping".
2. Tap and hold the SHIFT key
3. Tap the `H` key and release it.
4. Release the SHIFT key.
5. Verify that an `H` is output and that the layer returns to `default` after step 4.
6. Repeat steps 2-5 five times.
7. Repeat steps 2-4 _quickly_ for at least 10 seconds.
    - Do not bother checking the result layer; the priority is to be quick here.
    - FAIL this test if any output text is replaced or removed during these repetitions.
    - FAIL this test if text (either `h` or `H`) is not output each time you press the `h` / `H` key.
8. When done with step 7's repetitions, tap the numeric (`123`) key and verify that the layer switches properly to keys with numbers in the top row and symbols in the middle two rows.

**TEST_FLICKS_BASIC**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "English...", then "Gesture Prototyping".
2. Tap and hold the `u` key.
3. Drag it in an northwest direction (up and to the left) for at least a key-width.
4. Release it.
5. Expected result: `ù`
6. Tap the `u` key again and drag it in a northeast direction for at least a key-width before releasing it.
7. Expected result: `ú` (does not replace the previous output)
8. Tap `u` again and drag north (straight up) in the same manner.
9. Expected result: `û`.

**TEST_FLICK_CORRECTION**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "English...", then "Gesture Prototyping".
2. Tap and hold the `u` key.
3. Drag it in an purely west direction (straight left).
4. Release it.
5. Expected result: `ù`
6. Tap and hold the `u` key.
7. Drag it purely south (straight down).
8. Release it.
9. Expected result:  `u`.  The flick animation and key preview should disappear.
10. Tap and hold the `u` key again.
11. Drag it up just a few pixels / millimeters - a _very_ short distance - then release it.
12. Expected output: `u`.

**TEST_FLICK_LOCKING**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "English...", then "Gesture Prototyping".
2. Tap and hold the `u` key.
3. For the following steps, pay attention to the animated preview.
4. Drag it purely north (straight up) - you should see the `û` key in the preview area (assuming it's not blocked by your finger).
    - If the animation was "jumpy" - if there was a lack of motion at the same time you moved your finger, then a big "jump" to catch up - FAIL this test.
    - If any such "jumps" were directly tied to you moving your finger quickly, do _not_ fail this test. 
5. Return your finger to its original position; the 'u' key should slide back into place. 
    - Note:  "original position" is kind of important; if you return slightly to the left of the original position, the next step may not work well.
6. Now, drag your finger in a up-and-right motion:  you should also see a preview for `ú` as you do so.
    - Again, if the animation felt "jumpy" and not well connected to your input, FAIL this test.
7. While the `ú` is visible and in the center, lift your finger and end the gesture.
8. Verify that a `ú` is output.
9. Repeat steps 2 through 6 for the `a` key; you should see similar previews and text there.  (`â`, `á`)

**TEST_MODIPRESS_MULTITAP_FLICK_PREVIEW**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "English...", then "Gesture Prototyping".
2. Double-tap and hold the SHIFT key, keeping it stationary for all following steps.
3. Tap and hold the `U` key.
4. Verify that the key preview is visible.
5. Drag the finger on `U` purely north (straight up) - you should see the `Û` key in the preview area (assuming it's not blocked by your finger).
6. Verify that the flick-animation in the preview scrolls to a `Û`.
7. With the `Û` fully in-view, release the flick.
8. Verify that a `Û` is output.
9. Repeat steps 2 through 6 quickly three times, including the "verify" steps.

**TEST_FLICK_DURING_MODIPRESS**:  Using the standard "Test unminified KeymanWeb" test page from a mobile device...

1. Select "English...", then "Gesture Prototyping".
2. Tap and hold the SHIFT key, keeping it stationary for all following steps.
3. Tap and hold the `U` key.
4. For the following steps, pay attention to the animated preview.
5. Drag it purely north (straight up) - you should see the `Û` key in the preview area (assuming it's not blocked by your finger).
    - If the animation was "jumpy" - if there was a lack of motion at the same time you moved your finger, then a big "jump" to catch up - FAIL this test.
    - If any such "jumps" were directly tied to you moving your finger quickly, do _not_ fail this test.  
6. Release the SHIFT key (held since step 2).
7. Verify that the shift layer (all uppercase) remains visible and that the flick animation is still active.
8. Return your finger to its original position; the 'U' key should slide back into place. 
12. Now, drag your finger in a up-and-right motion:  you should also see a preview for `Ú` as you do so.
    - Again, if the animation felt "jumpy" and not well connected to your input, FAIL this test.
13. While the `Ú` is visible and in the center, tap on the `Q` key.
14. Verify that the keyboard returns to the default (lowercase) layer and that either `Ú` or `Úq` is output.
    - `Úq` is ideal, but if your `a` tap is quick, you may only get the `Ú`.

I will note that sometimes, when doing the input quickly, I get only `Ú` (and not the `q`).  Not sure exactly why that is, but I felt it may be worth noting... and worth a future investigation.  It is likely typing-speed related, since there's a fair bit of `await`-`async` there.

mcdurdin commented on common/web/gesture-recognizer/src/engine/headless/touchpointCoordinator.ts:
```diff
@@ -109,55 +119,84 @@ export class TouchpointCoordinator<HoveredItemType, StateToken=any> extends Even
+    // For multitouch gestures, only report the gesture **once**.
+    const sourceIDs = selection.matcher.allSourceIds;
+    for(let sequence of this._activeGestures) {
+      if(!!sequence.allSourceIds.find((id1) => !!sourceIDs.find((id2) => id1 == id2))) {
```
O(m*n)? Are these arrays short?

jahorton replied:
A person only has ten fingers, and I'd be quite impressed if someone actually managed to use each of them independently for concurrent gestures.

Secondly, with the way Web currently works, I believe the maximum count for concurrent gestures is 2; a modipress and whatever the current input is on the modipress layer.  Starting a third will immediately resolve the previously-current input (simple tap, flick, multitap) or be blocked (longpress).  There's a _chance_ of it actually being 3 - I'm not 100% sure that a held backspace will be auto-cancelled on new input.  But yeah, that should be it.

This _does_ depend on proper state management of the 'active gestures' array.  If errors happen that prevent removal of finished gesture objects from that array, then over time it may become "less short" than intended.

mcdurdin replied:
That's what I figured but helpful info on the potential edge cases, thanks!

bharanidharanj commented:
# Test Results

* **TEST_10_KEY_ROTA (PASSED):** Tested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Opened "English - Classic 10-key" keyboard. 2. Verified that all of the keys numbered from 1 to 9 aside from 7 have three or more lowercase letters as key hints. 3. Followed steps 3 and 4 and verified that the expected output is 8a. 4. Tapped the 8 key continuously for at least 2 seconds and verified that it output different combination of letters and numbers. Stopped tapping when the new character is a capital letter. Verified that the key hints are now capitalized. 5. Double-tapped the 9 key once. A 'D' key appeared in the text area. 6. Repeatedly tapping 9 for at least 5 seconds and verified that the key hits change from lowercase letter to uppercase letter or uppercase letter to lowercase letter. 

..1to9 numbers with key hints

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/68e331ed-7d25-4fe1-9c8d-b3e798a312be">


..Tap the 8 key twice in quick succession

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/29f2561d-7cdc-4a72-85fd-16be992a562b">


..Tap the 8 key continuously for at least 2 seconds

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/8e984aa3-07ff-4de1-8e0f-3342c6544cd9">


..Repeatedly tap 9 for at least five seconds..

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/8c0eb0cc-f6da-4b41-9406-c5d2b0442034">

bharanidharanj commented:
# Test Results

* **TEST_10_KEY_DIACRITICS (PASSED):** Tested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "English - Diacritic 10-key Rota" keyboard. 2. Tapped the ◌̀ key. Then, double tapped the 8 Key. Verified that the expected result à appear on the text area. 3. Triple-tapped the ◌̀  key. Triped tapped the 9 key. Verified that the àê letters appeared on the screen. 4. Tapped the '◌̀' key. Then, tapped the 4 key once. Hit backspace. Double-tapped the 8 key showed the expected result àềa. 

..result 1

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/3f41acf3-7be3-490e-a981-551fd02e6689">

..result 2

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/82deacf3-6686-4ae0-8712-6a5bf0a3e7c4">

..result 3

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/f03dc03c-30c5-4a78-8df7-437abdb22bdd">

bharanidharanj commented:
# Test Results

* **TEST_CUSTOM_MULTITAP_MODIFIER (FAILED):** Tested using the standard "Test unminified KeymanWeb" test page from the mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "English - Gesture Prototyping" keyboard. 2. Single-tapped the ◌́  key. Noticed that the number of letters, including all of the vowels are changed to versions using the wrong output of the accent mark. Seems to be an issue. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/b2e5d8f2-1b54-42ce-980a-f9429f3cada2">

bharanidharanj commented:
# Test Results

* **TEST_BASIC_SIMPLE_SHIFT (PASSED):** Tested using the standard "Test unminified KeymanWeb" test page from the mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Norwegian" keyboard. 2. Tapped the Spacebar key. Verified that the keyboard was in the default (lowercase) layer after this. 3. Tapped the shift key once. Verified that the keyboard changes to, and stays on, the shift layer. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/8e11ae0c-c148-4012-9cfd-34bf5c2b590d">



* **TEST_BASIC_MODIPRESS (FAILED):** Tested using the standard "Test unminified KeymanWeb" test page from the mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Norwegian - EuroLatin (SIL)" keyboard. 2. Tapped the Spacebar key. Verified that the keyboard was in the default (lowercase) layer after this. 3. Tapped and hold the shift key. Verified that the keyboard changes to the shift layer. 4. Tapped the A key. Verified that it produces an A in the text area. Release the shift key. Verified that the keyboard is changed to the default layer as a result of releasing the shift key. Repeated steps 3 and 4 quickly and noticed that the shift layer was stuck at one point. Pressing on the Shift key does not do anything on the OSK. Seems to be an issue.

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/e3a33faf-e9e3-4c23-8f04-8e68c6cc0088">

bharanidharanj commented:
# Test Results

* **TEST_BASIC_MODIPRESS_HOLD (PASSED):** Tested using the standard "Test unminified KeymanWeb" test page from the mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Norwegian" keyboard. 2. Tapped the Spacebar key. Verified that the keyboard was in the default (lowercase) layer after this. 3. Tapped and hold the shift key. Verified that the keyboard is changed to the shift layer. 4. After one second, released the shift key. Verified that the keyboard is changed to the default layer.  

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/ab15bb71-e31d-44b0-96b6-b80546dc463c">


* **TEST_NUMERIC_FROM_SHIFT (PASSED):** Tested using the standard "Test unminified KeymanWeb" test page from the mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Norwegian - EuroLatin (SIL)" keyboard. 2. Tapped the Spacebar key. Verified that the keyboard was in the default (lowercase) layer after this. 3. Tapped the shift key, to see the shift layer active on the keyboard. 4. Tapped and hold the numeric key (123) underneath the shift key. Verified that the keyboard is changed to the numeric layer. 5. Longpressed then selected the  ‱ subkey. Verified that ‱ is emitted as text. 6. Release the the numeric key. Verified that the keyboard changes back to the shift layer. 7. Tapped the numeric key, releasing it immediately. 8. Repeated step 5 and verified that the keyboard automatically changes back to the default layer. 

..subkey text 1

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/19f17586-d84b-46ec-a6f1-765eb706c2bd">


..subkey text 2

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/13d411f6-8282-4682-ab21-c07ef970f8d9">

bharanidharanj commented:
# Test Results

* **TEST_DELAYED_SUBKEY (FAILED):** Tested using the standard "Test unminified KeymanWeb" test page from the mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Norwegian - EuroLatin (SIL)" keyboard. 2. Tapped the Spacebar key. Verified that the keyboard was in the default (lowercase) layer after this. 3. Tapped the shift key, to activate the shift layer. 4. Tapped and hold the numeric key 9123) underneath the shift key. Verified that the keyboard changes to a numeric layer. 5. Longpressed the % key, without selecting a subkey yet. 6. Released the numeric key. Verified that the keyboard changes remain on the numeric layer, with the subkey menu still displayed. 6. Selected the  ‱ subkey. I noticed that the layer has changed back to the default (lowercase) layer after getting the output ‱ on the screen. Seems to be an issue. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/53e3729e-1ae9-40ee-8e79-f133cb1bff82">

bharanidharanj commented:
# Test Results

* **TEST_DOUBLETAP_CAPS (FAILED):** Tested using the standard "Test unminified KeymanWeb" test page from the mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. After running the 4th step (ie., Without touching the shift key at any point, type CAPS), the shift stuck with this step. 2. Single-tapped the shift key does not return to the base layer. Seems to be an issue.

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/61bedad0-010c-40b4-8a3b-ab2c88c1345a">

bharanidharanj commented:
* **TEST_FLICKS_BASIC (FAILED):** Tested using the standard "Test unminified KeymanWeb" test page from the mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. After running the 4th step it is showing the wrong output on the text screen. seems to be an issue.

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/ef819d39-b70c-4269-8a37-0274e87e8408">

bharanidharanj commented:
# Test Results

* **TEST_FLICK_CORRECTION (PASSED):** Tested using the standard "Test unminified KeymanWeb" test page from the mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Followed the steps that are mentioned in the Test_Flick_Correction and it is working as expected. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/079b776d-1898-413b-b7a2-deffffd19ec1">


* **TEST_FLICK_LOCKING (PASSED):** Tested using the standard "Test unminified KeymanWeb" test page from the mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Followed the steps that are mentioned in the Test_Flick_Locking and it is working as expected. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/da2aeb0d-5bfe-4b0d-a958-9d1e288b0d13">

bharanidharanj commented:
# Test Results

* **TEST_FLICK_DURING_MODIPRESS (FAILED):** Tested using the standard "Test unminified KeymanWeb" test page from the mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. This test failed in Step 7 because the Shift layer (all uppercase) changed to the default (all lowercase) layer. (ie., after releasing the shift key) Seems to be an issue. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/53357bc5-e260-406c-8e75-e9ab71b2625d">

jahorton commented:
TEST_ALTERNATING_SHIFT_AND_KEY: FAIL (please wait for the all-in-one retest request)

TEST_CUSTOM_MULTITAP_MODIFIER: FAIL  (please wait for the all-in one retest request)

TEST_APP_10_KEY_DIACRITICS : FAIL  (please wait for the all-in one retest request)

jahorton commented:
@keymanapp-test-bot retest all

Just pushed a _bunch_ of fixes and such for the various issues we've been noting - things _should_ work a lot better now, assuming the build goes through.

Since there _have_ been a lot of changes, it's best to be sure I didn't break something that had been working before.

bharanidharanj commented:
# Test Results

* **TEST_10_KEY_ROTA (PASSED):** Tested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Opened "English - Classic 10-key" keyboard. 2. Verified that all of the keys numbered from 1 to 9 aside from 7 have three or more lowercase letters as key hints. 3. Followed steps 3 and 4 and verified that the expected output is 8a. 4. Tapped the 8 key continuously for at least 2 seconds and verified that it output different combination of letters and numbers. Stopped tapping when the new character is a capital letter. Verified that the key hints are now capitalized. 5. Double-tapped the 9 key once. A 'D' key appeared in the text area. 6. Repeatedly tapping 9 for at least 5 seconds and verified that the key hits change from lowercase letter to uppercase letter or uppercase letter to lowercase letter.

..Expected Output 8a

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/81b0107b-b8b3-4320-926b-f6ed9f207929">


..Typing 8 key continuously 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/97b5d469-3975-4ac7-920e-3b775d5c564e">

..Double-tap the 9 key once

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/bf3c9360-fdb2-4839-ba1b-856fecf343a6">


..Repeatedly tap 9 key for at least five seconds

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/8be939ae-70bd-44ee-bff8-890f7e61f6d8">

bharanidharanj commented:
Test Results

* **TEST_10_KEY_DIACRITICS (FAILED):** Tested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "English - Diacritic 10-key Rota" keyboard. 2. Tapped the ◌̀ key. Then, double-tapped the 8 Key. Verified that the expected result à appears on the text area. 3. Triple-tapped the ◌̀ key. Triped tapped the 9 key. Verified that the àê letters appeared on the screen. 4. Tapped the '◌̀' key. Then, tapped the 4 key once. Hit backspace. Double-tapped the 8 key showing the wrong output. 

..Double tap 8 key

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/c270bc42-e672-4863-b598-0e5c64bb8721">

..Triple-tap the 9 key

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/b4115461-fd57-45ca-a542-ab04117e2d7d">

..Double-tap the 8 key

 <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/2fcca831-78af-4d0b-a21f-056165303a92">

jahorton commented:
Oh right, we changed that behavior with #9944.  Backspace doesn't nuke the deadkeys anymore... which means I need to rewrite that test to use a different deadkey-erasing behavior.

jahorton commented:
@keymanapp-test-bot retest TEST_10_KEY_DIACRITICS 

@bharanidharanj That failed test is on me; I forgot that we changed a related behavior with #9944.  I've updated the test instructions accordingly.

Turns out that I discovered a bug on `master` while doing so; for the sake of user testing, I've gone ahead and fixed it here.  It'd require a retooling of the test keyboard and heavier retooling of the instructions otherwise.

----

Also, something probably worth note/making an issue:  TEST_APP_10_KEY_DIACRITICS will fail if performed on an iOS device if the `7` key step is removed... if the text area is empty before starting the test.  My intuition tells me that this is due to one of the iOS quirks we noticed a while back:  the iOS equivalent to a text-services framework signals a context reset any time the active text-input receiver has its text cleared.  Well... multitapping on a context-initial diacritic-output-only key probably isn't helping things there, possibly causing a context-desync between the app and the keyboard engine.  That doesn't seem _entirely_ right, but it's probably quite related.

As it appears to be due to something iOS-specific and does not occur on Android, I think this issue should be documented and pursued separately.

On Android, the context-initial diacritic doesn't seem to get displayed properly in the emulator setup I use most frequently, but it _is_ still present and _does_ still combine properly according to keyboard rules.  (Notably, placeholder text disappears when it should be visible, indicating the presence of text... even if it - the isolated diacritic - is not being displayed properly.)  That'd be more of a rendering issue within the app and/or on Android, I guess.

jahorton commented:
> Also, something probably worth note/making an issue: TEST_APP_10_KEY_DIACRITICS will fail if performed on an iOS device if the `7` key step is removed... if the text area is empty before starting the test. My intuition tells me that this is due to one of the iOS quirks we noticed a while back: the iOS equivalent to a text-services framework signals a context reset any time the active text-input receiver has its text cleared. Well... multitapping on a context-initial diacritic-output-only key probably isn't helping things there, possibly causing a context-desync between the app and the keyboard engine. That doesn't seem _entirely_ right, but it's probably quite related.
> 
> As it appears to be due to something iOS-specific and does not occur on Android, I think this issue should be documented and pursued separately.

Eh, why not.

Did a little investigation:  our old enemy, U+200E (https://unicode-explorer.com/c/200E) has returned to haunt us.  It's being automatically inserted at the start of the context whenever a diacritic is added in that position or directly after a different diacritic meeting this compound condition (e.g., it can chain).

This is occurring via `setKeymanVal` calls to the ios-host.js script:

https://github.com/keymanapp/keyman/blob/6b8a06a52faa16d3f1d3501ff8ac5ee66f1535fa/ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/ios-host.js#L292-L295

This has been verified via breakpoint when performing a repro for the conditions described above - that if-condition will trigger, resulting in, among other things, a deadkey-wipe.

The Swift-side interface that calls this:

https://github.com/keymanapp/keyman/blob/6b8a06a52faa16d3f1d3501ff8ac5ee66f1535fa/ios/engine/KMEI/KeymanEngine/Classes/Keyboard/KeymanWebViewController.swift#L231-L251

Well, huh.  That _should_ handle it, right?  Well... here's the thing...

![Screen Shot 2023-11-22 at 10 27 52 AM](https://github.com/keymanapp/keyman/assets/25213402/f041ab15-2e9f-47e6-b145-ec5b9ecb96ea)

![Screen Shot 2023-11-22 at 10 26 16 AM](https://github.com/keymanapp/keyman/assets/25213402/ca16c7f2-06f7-49b1-b7e8-6f75234365c5)

So... there's no match, the byte-header doesn't get filtered, do not pass Go, do not collect $200, etc.  Of course, that _did_ work before... so a recent-ish version of iOS must have changed how the byte-header gets encoded.

Also, I'd noticed weirdness with predictive-text even on `master` - how an extra backspace was needed after reaching cleared text at the start of context for predictions to work right.  It's definitely the same thing - our pred-text engine doesn't filter out byte-order marks either.

jahorton commented:
Given the investigation and related behaviors I've noted, it's a bug on the current `master` and thus should be handled separately.

jahorton commented:
And, of course, fixing the fact that context-resets weren't resetting deadkeys (a la #10039)... caused me to run into a different context-sync issue, at least on iOS.  Now checking if it also happens on Android.

Core issue there:  there are two text updates during a multitap:  'restore original context' and then 'apply new text effects'.  If text is updated from the app based on the first and not the second, while the JS side has done both, that causes a context desync.

jahorton commented:
As that last comment is something only relevant for `feature-gestures` - because of multitaps - I went ahead and fixed that here.

bharanidharanj commented:
# Test Results

* **TEST_10_KEY_DIACRITICS (PASSED):** Tested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "English - Diacritic 10-key Rota" keyboard. 2. Tapped the ◌̀ key. Then, double-tapped the 8 Key. Verified that the expected result à appears in the text area. 3. Triple-tapped the ◌̀ key. Triped tapped the 9 key. Verified that the àê letters appeared on the screen. 4. Tapped the '◌̀' key. Then, tapped the 4 key once. Hit backspace. Double-tapped the 8 key showed the expected result àềa.

..Double-tap  8 key

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/a4a0542c-c38a-45ae-a486-111d81a786af">

..Triple-tap  9 key

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/b267b595-27ef-48e3-8127-19039e896487">

..Double-tap the 8 key after the change

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/290ba8ff-6a3c-440a-87a2-219dddabc4d5">

bharanidharanj commented:
# Test Results

* **TEST_APP_10_KEY_DIACRITICS (PASSED):** Tested using the Keyman for Android test artifact.. from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Installed the attached Keyman build 17.0.209-alpha-test-9973 in the device. 2. Downloaded and then installed diacritic_rota.kmp file in Keyman App. 3. Tapped the 7 key twice. Verified that 77 appears on the screen. 4. After running step 5, able to see 77à as a result. 5. After running step 7, able to see the 77àê as a result. 6. After running step 11, able to see 77àềa as a result. 

..Double-tap the 8 key

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/d435db5e-93dd-4c66-a4e2-cc198e7508a4">

..Triple-tap the 9-key

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/7bbe434e-0d4d-4218-a0c4-befbeca378fc">

..Double-tap the 8th key

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/b7624c40-cd59-4fdd-91d5-8e3eb0528f28">

bharanidharanj commented:
# Test Results

* **TEST_BASIC_SIMPLE_SHIFT (PASSED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Norwegian" keyboard. 2. After running step 3, it is verified that the keyboard remains on the shift layer. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/5be95fa5-236f-4b44-ad23-4d80d7c0b040">


* **TEST_BASIC_MODIPRESS (PASSED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Norwegian" keyboard. 2. After running step 8, it is verified that the keyboard changed to the default layer. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/38e7bd7c-99db-4f55-9b89-b45cf056331b">


* **TEST_BASIC_MODIPRESS_HOLD (PASSED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Norwegian" keyboard. 2. After running step 6, it is verified that the keyboard changed to the default layer due to step 6.

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/93d53941-7906-45c5-9840-9782333a9f6a">

bharanidharanj commented:
# Test Results

* **TEST_DELAYED_SUBKEY (PASSED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Norwegian" keyboard. 2. Verified that the keyboard keyboard changed to the numeric layer after running step 4. 3. Also, verified that the keyboard changes remain on the numeric layer, with the subkey menu still displayed, after running step 8. 4. Verified that ‱ is emitted as text and that the layer has changed back to the shift layer. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/b29a7553-ea8f-4489-9b8f-399fa2be675a">

bharanidharanj commented:
# Test Results

* **TEST_DOUBLETAP_CAPS (PASSED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Norwegian" keyboard. 2. Verified that it is showing the expected outputs after running from 2 to 8. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/a2678737-740d-4374-98e9-647cc4125032">

bharanidharanj commented:
# Test Results

* **TEST_CUSTOM_MULTITAP_MODIFIER (FAILED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Gesture Prototyping" keyboard. 2. Noticed that after running step 4, it is showing the wrong output in the text screen.  

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/e3606ea5-289d-4512-bc2c-b9a0fd8551f1">

bharanidharanj commented:
# Test Results

* **TEST_ALTERNATING_SHIFT_AND_KEY (PASSED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Gesture Prototyping" keyboard. 2. Verified that an H is output and the keyboard layer returns to default layer after running step 4. 3. Verified that there is no deviations in the output after running step 7. 4. Verified that the layer switches properly to keys with numbers in the top row and symbols in the middle two rows, after clicking the 'num' key. 

..H is the output after running step 5

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/317aac00-c74b-4d4c-b98a-9fe3d2ede7f2">


..No deviation in the output after running step 7

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/cc84a871-391c-409d-99f4-cf3e56dddfbd">


..Numeric key appears after running step 8

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/69843d15-6184-4a23-9d76-bc7e3f7c35cb">

bharanidharanj commented:
# Test Results

* **TEST_FLICKS_BASIC (PASSED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Gesture Prototyping" keyboard. 2. After running step 4, it is verified that it outputs ù on the screen. 3. After running step 6, it is verified that it outputs ù on the screen. 4. After running step 8, it is verified that it outputs û on the screen.

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/d981ccf5-15e7-4a69-8466-5a72304f511f">

bharanidharanj commented:
# Test Results

* **TEST_FLICK_CORRECTION (FAILED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Gesture Prototyping" keyboard. 2. After running step 4, it is verified that it outputs ù on the screen. 3. After running step 8, it is observed that the u letter appears on the screen. seems to be an issue. 4. After running step 11, it is verified that it outputs u on the screen. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/e24ad0c3-66fe-425b-80cb-0897e62c5ff9">

jahorton commented:
> # Test Results
> 
>     * **TEST_CUSTOM_MULTITAP_MODIFIER (FAILED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Gesture Prototyping" keyboard. 2. Noticed that after running step 4, it is showing the wrong output in the text screen.
>        <img alt="" width="300" src="https://user-images.githubusercontent.com/19683143/284931323-e3606ea5-289d-4512-bc2c-b9a0fd8551f1.JPEG">

@keymanapp-test-bot retest TEST_CUSTOM_MULTITAP_MODIFIER 

@bharanidharanj Could you try again and verify that you _triple-tapped_ the key?  The results you show would be correct for a _double-tap_ of the layer-shifting key, which is pretty close.

jahorton commented:
> # Test Results
> 
> * **TEST_FLICK_CORRECTION (FAILED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Gesture Prototyping" keyboard. 2. After running step 4, it is verified that it outputs ù on the screen. 3. After running step 8, it is observed that the u letter appears on the screen. seems to be an issue. 4. After running step 11, it is verified that it outputs u on the screen.
>        <img alt="" width="300" src="https://user-images.githubusercontent.com/19683143/284956899-e24ad0c3-66fe-425b-80cb-0897e62c5ff9.JPEG">

**TEST_FLICK_CORRECTION (PASSED):** In which it turns out another test spec needed updating to match current behavior.  I think I wrote that test before the locked-flicks stuff was fully implemented; it makes sense for step 8 to emit that 'u', now that I inspect it more closely.

I've gone ahead and adjusted the test-spec accordingly.

bharanidharanj commented:
# Test Results

* **TEST_FLICK_LOCKING (PASSED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Gesture Prototyping" keyboard. 2. Tapped and hold the 'u' key. 3. Dragged it purely straight-up position, and I was able to see the letter û. Observed that there was no lack of motion after dragging it. 4. After moving back my finger to the original position, the 'u' key slides back into place. 5. Tapped and hold the 'u' key, then drag it in an up-and-right motion. I was able to see a preview of ú key. Observed that no animated preview happened. Lift my finger and end the gesture. Verified that a ú outputs on the screen.  6. Followed steps 2 through 6 for the 'a' key and I am able to see the similar previews and text there. Seems to be working fine. Thanks.


..u text animated preview

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/c3a9b543-e9d3-433d-ba93-d4d5e025da78">

..a text-animated preview

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/c7195e9f-dff6-4733-a796-d9d39a8619c5">

bharanidharanj commented:
# Test Results

* **TEST_MODIPRESS_MULTITAP_FLICK_PREVIEW (PASSED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Gesture Prototyping" keyboard. 2. Double tapped then hold the Shift key. Tapped and hold the U key. Verified that the key preview is visible. 3. Dragged the finger on U purely north (straight up) and I was able to see Û key in the preview area. 4. Verified that the flick-animation in the preview scrolls to a Û. 5. Release the flick, when the Û key is fully visible. Verified that Û is output. 6. Repeated steps 2 through 6 quickly three or four times and verified that the Û appears repeatedly as the output. Seems to be working fine. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/de0acaae-5cee-4b43-8fad-81d08ca6a713">

bharanidharanj commented:
# Test Results

* **TEST_FLICK_DURING_MODIPRESS (PASSED):** Retested using the standard "Test unminified KeymanWeb" test page from a mobile device (Samsung Galaxy A23 - Android 13) and here is my observation: 1. Selected "Gesture Prototyping" keyboard. 2. Tapped and hold the Shift key. Tapped and hold the U key. 3. Dragged it to straight-up position. I was able to see the Û key without "jumpy" in the preview area. Released the Shift key. Verified that the shift layer remains visible and that the flick animation is still active. 4. Returned the finger to its original position, able to see the U key slide back into place. 5. Dragged the finger in an up-and-right motion and I was able to see the Ú key in the preview. 6. While the Ú is visible in the center, tapped on the q key with the other finger. Verified that the keyboard returns to the default layer and Úq is output on the screen. Seems to be working fine. 

   <img width="300" src="https://github.com/keymanapp/keyman/assets/19683143/d5c4115a-bb81-4372-aeac-3d0921fa74e1">
